frontend/src/javascript/utils/codeUtils.js

```diff
@@ -59,37 +60,82 @@ export function monacoChangeToLegacyFormat(code, change) {
     return result;
 }
 
-function handleVarTrackerMove(change, varObj) {
+function isLegalVariableName(text) {
+    return /^[a-zA-Z_][a-zA-Z_0-9]*$/.test(text);
+}
+
+function isLegalVariableCharacter(char) {
+    let code = char.charCodeAt(0);
+    return (code > 47 && code < 58) || (code > 64 && code < 91) || (code > 96 && code < 123) || code == 95;
+}
+
+function isDigit(char) {
+    return char >= "0" && char <= "9";
+}
+
+function expandLeft(varObj, project) {
+    let i = varObj.start - 1;
+    while (i >= 0 && isLegalVariableCharacter(project.code[i])) i--;
+    i++;
+    while (i <= varObj.end && isDigit(project.code[i])) i++;
+    varObj.start = i;
+}
+
+function expandRight(varObj, project) {
+    let i = varObj.end;
+    while (i < project.code.length && isLegalVariableCharacter(project.code[i])) i++;
+    varObj.end = i;
+}
+
+function findFirstLegalVariableName(text) {
+    const regex = /[a-zA-Z_][a-zA-Z_0-9]*/g;
+    let match = regex.exec(text);
+    if (match == null) return null;
+    return { start: match.index, end: regex.lastIndex };
+}
+
+function handleVarTrackerMove(change, varObj, project) {
     if (varObj == null) return;
 
-    if (areIntervalsIntersect(varObj.start, varObj.end, change.start, change.end)) {
-        if (isSubinterval(varObj.start, varObj.end, change.start, change.end)) {
-            varObj.end += change.size;
-            if (varObj.end - varObj.start <= 0) return "Delete";
-            return "Rename";
-        }
-        return "Delete";
+    let originalPos = { start: varObj.start, end: varObj.end };
+
+    let newVarRange = applyChangeOnInterval(varObj.start, varObj.end, change);
+    if (newVarRange.end - newVarRange.start <= 0) return "Delete";
+    varObj.start = newVarRange.start;
+    varObj.end = newVarRange.end;
+
+    if (!areIntervalsIntersectOrTouching(originalPos.start, originalPos.end, change.start, change.end)) {
```

Comment:
Co do tego mam jeden komentarz - w (skrajnie dziwnym) przypadku, gdy mamy `123abc` i śledzimy to `abc`, a następnie dodamy litery przed cyfry i powstanie `xyz123abc` to teoretycznie jest to poprawna zmienna (i może nawet powinna być śledzona). Jednak zmiana wydarzyła się z dala od śledzonej zmiennej `abc`. Przypadek na tyle skrajny że celowo trochę go ignoruję. Jednak ten warunek zwalnia z dość ciężkich obliczeń przy każdej zmianie kodu na każdej śledzonej zmiennej.

Reply:
Tak na prawdę wystarczy dodać wewnątrz tego ifa `expandLeft(...)` i to załatwia temat. Może dołoży to trochę dodatkowych obliczeń, ale nie wydaje mi się, żeby to skutkowało na wydajność. Jak zmienimy sposób zaznaczania śledzonych zmiennych to można sprawdzić czy nie ma przez to jakiś ścinek przy edytowaniu kodu (aktualnie proces dodawania kilkudziesięciu-kilkuset śledzonych zmiennych wydaje się zbyt żmudny i czasochłonny). Gdyby się jednak okazało, że to powoduje jakieś lagi przy edycji kodu to zawsze można zrobić jakąś flagę, że ten expandLeft wykonujemy tylko dla pierwszej zmiennej po prawej stronie od zmiany.